src/ray/object_manager/plasma/object_lifecycle_manager.cc

```diff
@@ -0,0 +1,281 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "ray/object_manager/plasma/object_lifecycle_manager.h"
+#include "absl/time/clock.h"
+#include "ray/common/ray_config.h"
+
+namespace plasma {
+using namespace flatbuf;
+
+ObjectLifecycleManager::ObjectLifecycleManager(
+    IAllocator &allocator, ray::DeleteObjectCallback delete_object_callback)
+    : object_store_(std::make_unique<ObjectStore>(allocator)),
+      eviction_policy_(std::make_unique<EvictionPolicy>(*object_store_, allocator)),
+      delete_object_callback_(delete_object_callback),
+      eargerly_deletion_objects_(),
+      num_bytes_in_use_(0) {}
+
+std::pair<const LocalObject *, flatbuf::PlasmaError> ObjectLifecycleManager::CreateObject(
+    const ray::ObjectInfo &object_info, plasma::flatbuf::ObjectSource source,
+    bool fallback_allocator) {
+  RAY_LOG(DEBUG) << "attempting to create object " << object_info.object_id << " size "
+                 << object_info.data_size;
+  if (object_store_->GetObject(object_info.object_id) != nullptr) {
+    return {nullptr, PlasmaError::ObjectExists};
+  }
+  auto entry = CreateObjectInternal(object_info, source, fallback_allocator);
+
+  if (entry == nullptr) {
+    return {nullptr, PlasmaError::OutOfMemory};
+  }
+  eviction_policy_->ObjectCreated(object_info.object_id, /*is_create=*/true);
+  return {entry, PlasmaError::OK};
+}
+
+const LocalObject *ObjectLifecycleManager::GetObject(const ObjectID &object_id) const {
+  return object_store_->GetObject(object_id);
+}
+
+const LocalObject *ObjectLifecycleManager::SealObject(const ObjectID &object_id) {
+  // TODO(scv119): should we check delete object from eargerly_deletion_objects_?
+  return object_store_->SealObject(object_id);
+}
+
+bool ObjectLifecycleManager::AbortObject(const ObjectID &object_id) {
```

Comment:
Does this bool return type make sense? It is unclear to me what's the right action to do when this returns false because there are 2 possible causes. Why don't we return error type instead? For example, in the upper layer, if abort failed because object is already sealed, it wants to call the DeleteObject, but it is not possible now